File: packages/ruby-sdk/lib/strait/composition/checkpoint_resume.rb
# frozen_string_literal: true

module Strait
  module Composition
    # Wraps a block with checkpoint/resume state management.
    #
    # @param ctx [Strait::Authoring::RunContext] The run context
    # @param last_checkpoint [Hash, nil] Restored state from prior run
    # @param initial_state [Hash] Default state if no checkpoint
    # @param checkpoint_interval [Integer] How often to checkpoint (default 1)
    # @yield [state, update_state] The user function
    # @return The result of the block
    def self.with_checkpoint_resume(ctx, last_checkpoint, initial_state:, checkpoint_interval: 1, &block)
      current_state = last_checkpoint || initial_state
      step_count = 0

      update_state = ->(new_state) {
        current_state = new_state
        step_count += 1
        if (step_count % checkpoint_interval).zero? && ctx.checkpoint
          begin
            ctx.checkpoint.call(current_state)
          rescue StandardError
            # fire-and-forget
          end
        end
      }

      result = block.call(current_state, update_state)

      # Final checkpoint
      ctx.checkpoint&.call(current_state)

      result
    end
  end
end

File: packages/go-sdk/composition/checkpoint_resume.go
package composition

import (
	"github.com/strait-dev/go-sdk/authoring"
)

// CheckpointResumeOptions configures checkpoint resume behavior.
type CheckpointResumeOptions struct {
	InitialState       map[string]any
	CheckpointInterval int // default 1
}

// WithCheckpointResume wraps a function with checkpoint/resume state management.
func WithCheckpointResume[T any](
	ctx authoring.RunContext,
	lastCheckpoint map[string]any,
	fn func(state map[string]any, updateState func(map[string]any)) (T, error),
	opts CheckpointResumeOptions,
) (T, error) {
	currentState := opts.InitialState
	if lastCheckpoint != nil {
		currentState = lastCheckpoint
	}

	interval := opts.CheckpointInterval
	if interval <= 0 {
		interval = 1
	}
	stepCount := 0

	updateState := func(newState map[string]any) {
		currentState = newState
		stepCount++
		if stepCount%interval == 0 && ctx.Checkpoint != nil {
			_ = ctx.Checkpoint(currentState)
		}
	}

	result, err := fn(currentState, updateState)
	if err != nil {
		var zero T
		return zero, err
	}

	// Final checkpoint
	if ctx.Checkpoint != nil {
		_ = ctx.Checkpoint(currentState)
	}

	return result, nil
}
